File: templates/problems/18_bit_manipulation/02_num_one_bits/typescript/solution.ts
/**
 * Count set bits using n & (n-1) trick. O(k) time where k = number of 1 bits, O(1) space.
 */
export function solveHammingWeight(n: number): number {
    let count = 0;
    let nn = n;
    while (nn !== 0) {
        nn &= nn - 1;
        count++;
    }
    return count;
}

File: templates/problems/18_bit_manipulation/02_num_one_bits/go/solution.go
package num_one_bits

// solveHammingWeight counts set bits using n & (n-1) trick. O(k) time where k = number of 1 bits, O(1) space.
func solveHammingWeight(n uint32) int {
	count := 0
	for n != 0 {
		n &= n - 1
		count++
	}
	return count
}
